core/src/main/java/hudson/model/Queue.java

```diff
@@ -1191,6 +1191,20 @@ private CauseOfBlockage getCauseOfBlockage(Item i) {
                 return causeOfBlockage;
         }
 
+        if(!(i instanceof BuildableItem)) {
+            // Make sure we don't queue two tasks of the same project to be built
+            // unless that project allows concurrent builds. Once item is buildable it's ok.
+            //
+            // This check should never pass. And must be remove once we can completely rely on `getCauseOfBlockage`.
```

Comment:
Unclear whether we can now “completely rely on `getCauseOfBlockage`”. But this code just seems to be a reworking of some much older logic which I find hard to follow, from 84bca061dc4f330d17999bb15ee4b40cac227fd9 & 3216dface66707a1b069e51912513ef02beac79c. Perhaps this has been obsolete since bc66388f8550e748f898164711e352e6eb5da996? System dates to c04e63d9f0c0b1a3b0683e64838c685cdddcd265.